### PR TITLE
Today Extension

### PR DESCRIPTION
**Pull request by ywsang**
Implement the Today Extension (#147). 

Includes the ViewController for the Today Extension, different cells for the Today Extension, and a helper file for retrieving data from User Defaults. 

(Need to squash commits still.)

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -46,5 +46,19 @@ target 'TCAT' do
...
+
 end
+
+target 'Today Extension' do
+    # UI Frameworks	
+    pod 'SnapKit'
+
+    # Networking + Data
+    pod 'TRON', '~> 4.1.2'
+    pod 'SwiftyJSON', '~> 4.0'
```
Comment (mattbarker016) [inline Podfile]: Are we still using this? We should (try to) be on Codable now

File: TCAT/AppDelegate.swift [hunk 2]
```diff
@@ -16,11 +16,14 @@ import Crashlytics
...
 
+// This is used for app-specific preferences
+let userDefaults = UserDefaults.standard
+
 @UIApplicationMain
 class AppDelegate: UIResponder, UIApplicationDelegate {
 
     var window: UIWindow?
-    let userDefaults = UserDefaults.standard
+//    let userDefaults = UserDefaults.standard
```
Comment (Omarrasheed) [inline TCAT/AppDelegate.swift]: Nitpick: j delete the line I guess unless we still need it

File: TCAT/AppDelegate.swift [hunk 3]
```diff
@@ -50,7 +53,11 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
         JSONFileManager.shared.deleteAllJSONs()
 
         for (key, defaultValue) in userDataInits {
-            if userDefaults.value(forKey: key) == nil {
+            if key == Constants.UserDefaults.favorites {
```
Comment (mattbarker016) [inline TCAT/AppDelegate.swift]: Why was this added? There doesn't seem to be any difference

File: TCAT/Model/BusStop.swift [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+//  TCAT
+//
+//  Created by Austin Astorga on 3/26/17.
+//  Copyright © 2017 cuappdev. All rights reserved.
+//
+
+import UIKit
+import CoreLocation
+
+@objc(BusStop) class BusStop: Place, CoordinateAcceptor {
```
Comment (mattbarker016) [inline TCAT/Model/BusStop.swift]: This model has been deprecated. It's now bundled in `LegacyModels.swift` as we converted away from it. Should this now be removed?

File: TCAT/Model/Direction.swift [hunk 5]
```diff
@@ -51,7 +51,7 @@ class Direction: NSObject, NSCopying, Codable {
     /// The starting time (UTC) associated with the direction. Format: `"yyyy-MM-dd'T'HH:mm:ssZZZZ"`
     var startTime: Date
     
-    /// The starting time (UTC) associated with the direction Format: `"yyyy-MM-dd'T'HH:mm:ssZZZZ"`
+    /// The ending time (UTC) associated with the direction Format: `"yyyy-MM-dd'T'HH:mm:ssZZZZ"`
```
Comment (mattbarker016) [inline TCAT/Model/Direction.swift:54]: 😍

File: TCAT/Model/Direction.swift [hunk 6]
```diff
@@ -63,7 +63,7 @@ class Direction: NSObject, NSCopying, Codable {
     /// The number representing the bus route.
     var routeNumber: Int = 0
     
-    /// An array of bus stop locations on the bus route, excluding the departure and arrival stop. Empty if `type != .depart`.
+    /// An array of bus stop locations on the bus route, including the departure and arrival stop. Empty if `type != .depart`.
```
Comment (mattbarker016) [inline TCAT/Model/Direction.swift:66]: 😍

File: TCAT/Model/Place.swift [hunk 7]
```diff
@@ -12,7 +12,7 @@ enum PlaceType: String, Codable {
     case busStop, googlePlace, unknown
 }
 
-class Place: NSObject, Codable {
+@objc(Place) class Place: NSObject, Codable {
```
Comment (mattbarker016) [inline TCAT/Model/Place.swift:15]: Is `@objc(Place)` needed for TodayExtension stuff?

File: TCAT/Model/PlaceResult.swift [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+//
+//  Created by Austin Astorga on 3/26/17.
+//  Copyright © 2017 cuappdev. All rights reserved.
+//
+import UIKit
+import TRON
+import SwiftyJSON
+import CoreLocation
+
+@objc(PlaceResult) class PlaceResult: Place, CoordinateAcceptor {
```
Comment (mattbarker016) [inline TCAT/Model/PlaceResult.swift]: Same note as `BusStop.swift` file / class

File: TCAT/Utilities/Endpoints.swift [hunk 9]
```diff
@@ -108,12 +107,12 @@ class Network {
             "destinationName"   :   end.name,
             "originName"        :   start.name
         ]
-        
+
         // Add unique identifier to request
-        if let uid = userDefaults.string(forKey: Constants.UserDefaults.uid) {
+        if let uid = sharedUserDefaults?.string(forKey: Constants.UserDefaults.uid) {
```
Comment (mattbarker016) [inline TCAT/Utilities/Endpoints.swift]: I just noticed, why is this now optional? I saw the global `let userDefaults = UserDefaults.standard` but do we then locally define `sharedUserDefaults`?

File: TCAT/Utilities/SearchTableViewHelpers.swift [hunk 10]
```diff
@@ -111,7 +119,11 @@ class SearchTableViewManager {
         
         do {
             let data = try encoder.encode(places)
-            userDefaults.set(data, forKey: key)
+            if key == Constants.UserDefaults.favorites {
```
Comment (mattbarker016) [inline TCAT/Utilities/SearchTableViewHelpers.swift:118]: Is there a reason we didn't just put everything within sharedUserDefaults for simplicity?

File: Today Extension/LoadingTableViewCell.swift [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+class LoadingTableViewCell: UITableViewCell {
+
+    let loadingIndicator = LoadingIndicator()
+    let verticalMargin: CGFloat = 20.0 // top & bottom margin
+
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        contentView.addSubview(loadingIndicator)    }
```
Comment (mattbarker016) [inline Today Extension/LoadingTableViewCell.swift]: watch that bracket!

File: Today Extension/LoadingTableViewCell.swift [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        loadingIndicator.snp.makeConstraints { (make) in
+            make.top.equalTo(verticalMargin)
+            make.bottom.equalToSuperview().inset(verticalMargin)
+            make.width.equalTo(40.0)
+
+            make.centerY.equalToSuperview()
```
Comment (mattbarker016) [inline Today Extension/LoadingTableViewCell.swift]: Nitpick: Can combine to `make.center.equalToSuperview()`

File: Today Extension/NoRoutesTableViewCell.swift [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+//  NoRoutesTableViewCell.swift
+//  Today Extension
+//
+//  Created by Yana Sang on 2/28/19.
+//  Copyright © 2019 cuappdev. All rights reserved.
+//
+
+import UIKit
+
+class NoRoutesCell: UITableViewCell {
```
Comment (mattbarker016) [inline Today Extension/NoRoutesTableViewCell.swift]: Nitpick: `NoRoutesAvailableTableViewCell` or `EmptyStateTableViewCell` is a bit better name to communicate what the cell does

File: Today Extension/NoRoutesTableViewCell.swift [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        noRoutesLabel.text = "Unable to Load Routes"
+
+        contentView.addSubview(noRoutesLabel)
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        noRoutesLabel.snp.makeConstraints { (make) in
+            make.centerY.equalToSuperview()
```
Comment (mattbarker016) [inline Today Extension/NoRoutesTableViewCell.swift]: Nitpick: Same as above

File: Today Extension/TodayExtensionCell.swift [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+    var busIcon: BusIcon?
+
+    // MARK: Spacing vars
+
+    let leftMargin: CGFloat =  12.0
+    let verticalMargin: CGFloat = 20.0 // top & bottom margin
+    let rightMargin: CGFloat = 16.0
+
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        departureLabel = UILabel()
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: Can you define these globally as defaults when you declare the variables?

File: Today Extension/TodayExtensionCell.swift [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        liveLabel = UILabel()
+        liveIndicatorView = LiveIndicator(size: .small, color: .clear)
+
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        contentView.addSubview(departureLabel)
+        contentView.addSubview(destinationLabel)
+        contentView.addSubview(liveLabel)
+        contentView.addSubview(liveIndicatorView)
+        if let busIcon = busIcon {
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: Why is this here?

File: Today Extension/TodayExtensionCell.swift [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+            contentView.addSubview(busIcon)
+        }
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        departureLabel.snp.makeConstraints { (make) in
+            make.top.equalTo(verticalMargin)
+            make.leading.equalTo(74.0)
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: Nitpick: should make a constant somewhere

File: Today Extension/TodayExtensionCell.swift [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+            busIcon.snp.makeConstraints { (make) in
+                make.top.equalTo(verticalMargin)
+                make.leading.equalTo(leftMargin)
+                make.height.equalTo(busIcon.intrinsicContentSize.height)
+                make.width.equalTo(busIcon.intrinsicContentSize.width)
+            }
+        }
+
+        destinationLabel.snp.makeConstraints { (make) in
+            make.top.equalTo(departureLabel.snp.bottom).offset(2)
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: Nitpick: Same as above

File: Today Extension/TodayExtensionCell.swift [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+            liveIndicatorView.snp.makeConstraints { (make) in
+                make.centerY.equalTo(liveLabel.snp.centerY)
+                make.leading.equalTo(liveLabel.snp.trailing).offset(8)
+                make.trailing.lessThanOrEqualTo(rightMargin)
+                make.width.equalTo(liveIndicatorView.intrinsicContentSize.width)
+            }
+        }
+    }
+
+    func setUpCell(route: Route?, destination: String) {
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: These types of statements (`self.route = route`) usually are in an init function, it might make more sense to make a custom init and set those variables there.

File: Today Extension/TodayExtensionCell.swift [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    func setUpNoRoute() {
+        showLiveElements = false
+        let noRouteLabel = UILabel()
+        noRouteLabel.font = .getFont(.regular, size: 14.0)
+        noRouteLabel.textColor = Colors.primaryText
+        noRouteLabel.textAlignment = .center
+        noRouteLabel.lineBreakMode = NSLineBreakMode.byTruncatingTail
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: Nitpick: Remove class

File: Today Extension/TodayExtensionCell.swift [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+        noRouteLabel.text = "No routes available to \(destinationName)."
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: This client-facing text should be in `Contents.swift`

File: Today Extension/TodayExtensionCell.swift [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+
+        contentView.addSubview(noRouteLabel)
+
+        noRouteLabel.snp.makeConstraints { (make) in
+            make.centerY.equalToSuperview()
+            make.centerX.equalToSuperview()
+            make.leading.equalToSuperview().inset(12.0)
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]:
```suggestion
            make.leading.trailing.equalToSuperview().inset(12.0)
```

Comment (mattbarker016) [reply Today Extension/TodayExtensionCell.swift]: And then remove below line

File: Today Extension/TodayExtensionCell.swift [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+            make.trailing.equalToSuperview().inset(12.0)
+        }
+    }
+
+    func setUpDepartureLabel() {
+        departureLabel.lineBreakMode = NSLineBreakMode.byTruncatingTail
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift]: Nitpicks: Remove Class name, move client-facing strings to Constants.swift

File: Today Extension/TodayExtensionCell.swift [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+        departureLabel.textColor = Colors.primaryText
+        if let route = route {
+            departureLabel.text = "To \(route.endName)"
+        }
+    }
+
+    private func getDelayState(fromDirection direction: Direction) -> DelayState {
+        let departTime = direction.startTime
+        if let delay = direction.delay {
+            let delayedDepartTime = departTime.addingTimeInterval(TimeInterval(delay))
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift:159]: Are we absolutely sure that delay could never be negative (e.g. bus is running early?) Might want to use logic to specifically check that that it's late if delatedDepartTime (departTime with delay) is later than departTime.

File: Today Extension/TodayExtensionCell.swift [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+    func setUpDestinationLabel() {
+        destinationLabel.font = .getFont(.regular, size: 16.0)
+        destinationLabel.textColor = Colors.secondaryText
+        destinationLabel.lineBreakMode = .byTruncatingTail
+
+        if let direction = busDirection {
+            let delayState = getDelayState(fromDirection: direction)
+
+            switch delayState {
+            case .late(date: let delayedDepartureTime):
```
Comment (mattbarker016) [inline Today Extension/TodayExtensionCell.swift:180]: Nitpick: Wonky spacing

File: Today Extension/TodayViewController.swift [hunk 26]
```diff
@@ -0,0 +1,226 @@
...
+
+        favorites = TodayExtensionManager.shared.retrieveFavoritesNames()
+        coordinates = TodayExtensionManager.shared.retrieveFavoritesCoordinates()
+
+        searchForRoutes()
+    }
+
+    func searchForRoutes() {
+        if
+            favorites.count > 0,
```
Comment (mattbarker016) [inline Today Extension/TodayViewController.swift]: As much as I like this `if` format, this would probably be better as one line

File: Today Extension/TodayViewController.swift [hunk 27]
```diff
@@ -0,0 +1,226 @@
...
+        let expanded = activeDisplayMode == .expanded
+        preferredContentSize = expanded ? CGSize(width: maxSize.width, height: 110.0 * CGFloat(favorites.count)) : maxSize
+
+        routesTable.reloadData()
+    }
+
+    func createConstraints() {
+        routesTable.snp.makeConstraints {(make) in
+            make.leading.trailing.equalToSuperview()
+            make.height.equalTo(110.0*5)
```
Comment (mattbarker016) [inline Today Extension/TodayViewController.swift]: Definitely make these constants, especially like `numberOfCells`. Wherever this should be used, this should be made a variable

File: Today Extension/TodayViewController.swift [hunk 28]
```diff
@@ -0,0 +1,226 @@
...
+        var nonNilFavorites = [String]()
+        var nonNilCoordinates = [String]()
+
+        var nilRoutes = [Route?]()
+        var nilFavorites = [String]()
+        var nilCoordinates = [String]()
+
+        for i in 0..<routes.count {
+
+            if (routes[i]?.directions.filter { $0.type == .depart })?.first != nil {
```
Comment (mattbarker016) [inline Today Extension/TodayViewController.swift]: `.sort` by `$0.type == .depart` to reduce all the variable usage

File: Today Extension/TodayViewController.swift [hunk 29]
```diff
@@ -0,0 +1,226 @@
...
+            // if: in compact mode, no favorites added, failed to load routes, or in loading state
+            return 1
+        } else {
+            return favorites.count
+        }
+
+    }
+
+    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
+        return 110.0
```
Comment (mattbarker016) [inline Today Extension/TodayViewController.swift]: Make this a constant along with the other time you use `110`

File: Today Extension/TodayViewController.swift [hunk 30]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        if (favorites.count != 0) {
+            if (routes.isEmpty) {
```
Comment (mattbarker016) [inline Today Extension/TodayViewController.swift]: Can chain these two conditionals together

File: Today Extension/TodayViewController.swift [hunk 31]
```diff
@@ -0,0 +1,226 @@
...
+        let cell = tableView.dequeueReusableCell(withIdentifier: Constants.TodayExtension.errorCellIdentifier, for: indexPath) as! TodayExtensionErrorCell
+        cell.boldLabel.text = Constants.TodayExtension.addFavorite
+        cell.mainLabel.text = Constants.TodayExtension.showTrips
+        cell.selectionStyle = .none
+        return cell
+    }
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+
+        var stringURL: String
```
Comment (mattbarker016) [inline Today Extension/TodayViewController.swift]: nittiest-of-picks: you can set `stringURL` to be "ithaca-transit://" (should also maybe be in Constants.swift), and then use `+=` later on

File: Today Extension/TodayExtensionCell.swift (hunk 23 above)
Comment (AAAstorga) [reply Today Extension/TodayExtensionCell.swift]: I disagree here. I think "To" does not belong in Constants file. It is too much abstraction

Comment (mattbarker016) [reply Today Extension/TodayExtensionCell.swift]: What about “No routes available”?

File: Today Extension/TodayViewController.swift (hunk 30 above)
Comment (ywsang) [reply Today Extension/TodayViewController.swift]: Was messing around with this, but wasn't sure of the best way to chain the conditionals for the rest of the code. Within `favorites.count != 0`, there's cases for `routes.isEmpty` and `!routes.IsEmpty`, which is why I separated the two conditionals in the beginning.

File: Today Extension/TodayViewController.swift (hunk 28 above)
Comment (ywsang) [reply Today Extension/TodayViewController.swift]: Changed up the logic a little bit for this function, but don't think it's possible to just use `.sort`. I need to sort the `favorites` array and `coordinates` array according to how `routes` is re-arranged, and I need to keep track of the index for that.

File: Today Extension/TodayExtensionCell.swift (hunk 24 above)
Comment (ywsang) [reply Today Extension/TodayExtensionCell.swift:159]: If the bus is running early, how do we want the cell to look? Just change the departure time to match the early time and report the bus as on time?

File: Podfile [hunk 32]
```diff
@@ -46,5 +46,18 @@ target 'TCAT' do
...
+    # UI Frameworks	
+    pod 'SnapKit'
+
+    # Networking + Data
+    pod 'TRON', '~> 4.1.2'
+    pod 'Alamofire', '~> 4.7'
+ 
+    # Analytics
+    pod 'Crashlytics', '~> 3.10'
+
```
Comment (young-k) [inline Podfile]: should this newline be here? @kevinchan159 @drewsdunne

File: TCAT/Supporting Files/Constants.swift [hunk 33]
```diff
@@ -232,8 +232,22 @@ struct Constants {
...
+        // cell identifiers
+        static let errorCellIdentifier = "errorCell"
+        static let loadingCellIdentifier = "loadingCell"
+        static let contentCellIdentifier = "todayExtensionCell"
+        
+        // cell strings
+        static let addFavorite = "Add a favorite "
+        static let showTrips = "to show favorite trips here."
+        static let unableToLoad = "Unable to Load Routes"
+        static let noRoutesAvailable = "No routes available to "
```
Comment (young-k) [inline TCAT/Supporting Files/Constants.swift:243]: should be alphabetized :)

File: Today Extension/TodayExtensionCell.swift [hunk 34]
```diff
@@ -0,0 +1,209 @@
...
+
+        if showLiveElements {
+            liveLabel.snp.makeConstraints { (make) in
+                make.bottom.equalToSuperview().inset(verticalMargin)
+                make.leading.equalTo(departureLabel)
+                make.height.equalTo(liveLabel.intrinsicContentSize.height)
+                make.width.equalTo(liveLabel.intrinsicContentSize.width)
+            }
+
+            liveIndicatorView.snp.makeConstraints { (make) in
```
Comment (young-k) [inline Today Extension/TodayExtensionCell.swift]: `make` instead of `(make)`

File: Today Extension/TodayExtensionErrorCell.swift [hunk 35]
```diff
@@ -0,0 +1,57 @@
...
+            if (boldLabel.text == nil) {
+                make.centerX.equalToSuperview()
+            } else {
+                make.leading.equalTo(boldLabel.snp.trailing)
+            }
+            make.width.equalTo(mainLabel.intrinsicContentSize.width)
+        }
+    }
+    
+    func setUpNoFavorites() {
```
Comment (young-k) [inline Today Extension/TodayExtensionErrorCell.swift]: ?

File: Today Extension/TodayExtensionCell.swift [hunk 36]
```diff
@@ -0,0 +1,209 @@
...
+
+            case .onTime(date: let departureTime):
+                liveLabel.textColor = Colors.liveGreen
+                let boardTime = Time.timeString(from: Date(), to: departureTime)
+                liveLabel.text = (boardTime == "0 min" ? "Board now" : "Board in \(boardTime)")
+                liveIndicatorView.setColor(to: Colors.liveGreen)
+
+                contentView.addSubview(liveLabel)
+                contentView.addSubview(liveIndicatorView)
+
```
Comment (young-k) [inline Today Extension/TodayExtensionCell.swift]: should we have newlines between the `case`s? @kevinchan159

**Review by young-k — COMMENTED** [review]
:very-nice-kevin:

File: Today Extension/TodayExtensionCell.swift (hunk 24 above)
Comment (mattbarker016) [reply Today Extension/TodayExtensionCell.swift:159]: Yeah!

**Comment by mattbarker016** [comment]
For the merge conflicts, we're leaning towards keeping master's (especially for `project.pbxproj` and `TCAT/TCAT.entitlements`). I think the other ones will also lean towards master (improved networking via schemes). I want you to work through this @ywsang, good practice! Reach out though for any help!

File: TCAT/Utilities/SearchTableViewHelpers.swift [hunk 37]
```diff
@@ -63,7 +62,17 @@ class SearchTableViewManager {
     }
 
     func retrievePlaces(for key: String) -> [Place] {
-        if
+        if (key == Constants.UserDefaults.favorites) {
+            if
+                let storedPlaces = sharedUserDefaults?.value(forKey: key) as? Data,
```
Comment (kevinchan159) [inline TCAT/Utilities/SearchTableViewHelpers.swift]: put 66 and 67 on same line

File: TCAT/Utilities/SearchTableViewHelpers.swift [hunk 38]
```diff
@@ -63,7 +62,17 @@ class SearchTableViewManager {
...
+                let favorites = try? decoder.decode([Place].self, from: storedPlaces)
+            {
```
Comment (kevinchan159) [inline TCAT/Utilities/SearchTableViewHelpers.swift]: put `{` at end of 68

File: TCAT/Utilities/SearchTableViewHelpers.swift [hunk 39]
```diff
@@ -63,7 +62,17 @@ class SearchTableViewManager {
...
+                return favorites
+            }
+            
+        }
+        
+        else if
```
Comment (kevinchan159) [inline TCAT/Utilities/SearchTableViewHelpers.swift]: put 75 at end of 73

File: Today Extension/LoadingTableViewCell.swift [hunk 40]
```diff
@@ -0,0 +1,37 @@
...
+    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
+        super.init(style: style, reuseIdentifier: reuseIdentifier)
+
+        contentView.addSubview(loadingIndicator)
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        loadingIndicator.snp.makeConstraints { make in
```
Comment (kevinchan159) [inline Today Extension/LoadingTableViewCell.swift:20]: should just put these constraints inside the `init` and not use `layoutSubviews()`

File: Today Extension/TodayExtensionCell.swift [hunk 41]
```diff
@@ -0,0 +1,205 @@
...
+        contentView.addSubview(departureLabel)
+        contentView.addSubview(destinationLabel)
+        contentView.addSubview(liveLabel)
+        contentView.addSubview(liveIndicatorView)
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        departureLabel.snp.makeConstraints { make in
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionCell.swift:40]: same comment about the constraint code

File: Today Extension/TodayExtensionErrorCell.swift [hunk 42]
```diff
@@ -0,0 +1,52 @@
...
+
+        boldLabel.font = .getFont(.medium, size: 14.0)
+        boldLabel.textColor = Colors.primaryText
+        contentView.addSubview(boldLabel)
+    }
+
+    override func layoutSubviews() {
+        super.layoutSubviews()
+
+        boldLabel.snp.makeConstraints { make in
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionErrorCell.swift]: same comment about constraints

File: Today Extension/TodayExtensionHelpers.swift [hunk 43]
```diff
@@ -0,0 +1,49 @@
...
+import Foundation
+
+class TodayExtensionManager {
+
+    let decoder = JSONDecoder()
+    static let shared = TodayExtensionManager()
+
+    func retrieveFavoritesNames() -> [String] {
+        if
+            let storedPlaces = sharedUserDefaults?.value(forKey: Constants.UserDefaults.favorites) as? Data,
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionHelpers.swift]: combine 17 + 18

File: Today Extension/TodayExtensionHelpers.swift [hunk 44]
```diff
@@ -0,0 +1,49 @@
...
+            let places = try? decoder.decode([Place].self, from: storedPlaces)
+        {
+            var favorites: [String] = []
+            for place in places {
+                favorites.append(place.name)
+            }
+            return favorites
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionHelpers.swift]: can simplify to `return places.map({ $0.name })`

File: Today Extension/TodayExtensionHelpers.swift [hunk 45]
```diff
@@ -0,0 +1,49 @@
...
+        } else {
+            return [String]()
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionHelpers.swift]: can just do `return []`

File: Today Extension/TodayExtensionHelpers.swift [hunk 46]
```diff
@@ -0,0 +1,49 @@
...
+        }
+    }
+
+    func retrieveFavoritesCoordinates() -> [String] {
+        if
+            let storedPlaces = sharedUserDefaults?.value(forKey: Constants.UserDefaults.favorites) as? Data,
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionHelpers.swift]: combine 32 + 33

File: Today Extension/TodayExtensionHelpers.swift [hunk 47]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var coordinates: [String] = []
+            for place in places {
+                if
+                    let lat = place.latitude,
+                    let long = place.longitude {
+                    coordinates.append("\(lat),\(long)")
+                }
+            }
+            return coordinates
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionHelpers.swift]: would rather do
```
return places.compactMap({
  if let lat = $0.latitude, let long = $0.longitude {
    return "\(lat),\(long)"
  }
  return nil
})
```

File: Today Extension/TodayExtensionHelpers.swift [hunk 48]
```diff
@@ -0,0 +1,49 @@
...
+        } else {
+            return [String]()
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionHelpers.swift]: `return []`

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
lgtm!

File: Today Extension/TodayExtensionCell.swift [hunk 49]
```diff
@@ -0,0 +1,220 @@
...
+                        make.bottom.equalToSuperview().inset(bottom)
+                        make.leading.equalToSuperview().inset(offset)
+                        make.width.equalTo(busIcon.intrinsicContentSize.width)
+                    }
+                }
+
+                setUpDepartureLabel()
+                setUpDestinationLabel()
+                setUpLiveElements()
+
```
Comment (kevinchan159) [inline Today Extension/TodayExtensionCell.swift]: could we add a `return` here and then just do `setupNoRoute()` once instead of twice

**Review by kevinchan159 — APPROVED** [review]
ITS TIME TO MERGE
